File: prototype-validation/journeyvalidation/page_validator.py
from bs4 import BeautifulSoup

from .model import Page, Pass, Fail


def validate_html(html: str, page: Page, label: str) -> list:
    soup = BeautifulSoup(html, "html.parser")
    results = [_validate_title(soup, page, label)]
    results.extend(_validate_elements(soup, page, label))
    return results


def _validate_title(soup, page: Page, label: str):
    h1 = soup.find("h1")
    if not h1:
        return Fail(f"{label} — title", page.title, "<no h1 found>")
    actual = h1.get_text(strip=True)
    if actual == page.title:
        return Pass(f"{label} — title matches")
    return Fail(f"{label} — title", f"'{page.title}'", f"'{actual}'")


def _validate_elements(soup, page: Page, label: str) -> list:
    validators = {
        "contentPage": _validate_content,
        "string": _validate_string,
        "datePage": _validate_date,
        "boolean": _validate_boolean,
        "radioButton": _validate_radio,
        "checkbox": _validate_checkbox,
        "multipleQuestionsPage": _validate_multiple_questions,
    }
    validator = validators.get(page.page_type)
    if validator is None:
        return [Fail(f"{label} — page type", "known page type", f"'{page.page_type}'")]
    return validator(soup, page, label)


def _validate_content(_soup, _page, label):
    return [Pass(f"{label} — content page (no form required)")]


def _validate_string(soup, _page, label):
    found = len(soup.find_all("input", {"type": "text"}))
    if found >= 1:
        return [Pass(f"{label} — has text input")]
    return [Fail(f"{label} — text input", "at least 1 <input type=text>", f"found {found}")]


def _validate_date(soup, _page, label):
    by_name = soup.find_all("input", attrs={"name": lambda n: n and any(
        k in n for k in ("day", "month", "year")
    )})
    by_type = soup.find_all("input", {"type": "date"})
    by_class = soup.find_all(class_="govuk-date-input")
    found = len(by_name) + len(by_type) + len(by_class)
    if found >= 1:
        return [Pass(f"{label} — has date fields")]
    return [Fail(f"{label} — date fields", "date input elements", f"found {found}")]


def _validate_boolean(soup, _page, label):
    found = len(soup.find_all("input", {"type": "radio"}))
    if found >= 2:
        return [Pass(f"{label} — has yes/no radios ({found} found)")]
    return [Fail(f"{label} — boolean radios", "at least 2 radio buttons", f"found {found}")]


def _validate_radio(soup, page: Page, label):
    radios = len(soup.find_all("input", {"type": "radio"}))
    label_texts = {lbl.get_text(strip=True) for lbl in soup.find_all("label")}

    results = []
    if radios >= len(page.options):
        results.append(Pass(f"{label} — radio count ({radios} found)"))
    else:
        results.append(
            Fail(f"{label} — radio count", f"{len(page.options)} radio buttons", f"found {radios}")
        )

    for opt in page.options:
        if any(opt in lt for lt in label_texts):
            results.append(Pass(f"{label} — option '{opt}' present"))
        else:
            results.append(Fail(f"{label} — option '{opt}'", f"label containing '{opt}'", "not found"))

    return results


def _validate_checkbox(soup, page: Page, label):
    found = len(soup.find_all("input", {"type": "checkbox"}))
    label_texts = {lbl.get_text(strip=True) for lbl in soup.find_all("label")}

    results = []
    if found >= len(page.options):
        results.append(Pass(f"{label} — checkbox count ({found} found)"))
    else:
        results.append(
            Fail(f"{label} — checkbox count", f"{len(page.options)} checkboxes", f"found {found}")
        )

    for opt in page.options:
        if any(opt in lt for lt in label_texts):
            results.append(Pass(f"{label} — option '{opt}' present"))
        else:
            results.append(Fail(f"{label} — option '{opt}'", f"label containing '{opt}'", "not found"))

    return results


def _validate_multiple_questions(soup, page: Page, label):
    all_text_elements = soup.find_all(["label", "legend"])
    all_text = {el.get_text(strip=True) for el in all_text_elements}

    results = []
    for q in page.questions:
        if any(q.question_title in t for t in all_text):
            results.append(Pass(f"{label} — question '{q.question_title}' present"))
        else:
            results.append(
                Fail(f"{label} — question", q.question_title, "not found on page")
            )
    return results

File: journey-validation/src/main/scala/journeyvalidation/PageValidator.scala
package journeyvalidation

import org.jsoup.Jsoup
import org.jsoup.nodes.Document
import scala.jdk.CollectionConverters._

object PageValidator {

  def validatePage(doc: Document, page: Page, pageLabel: String): List[ValidationResult] = {
    val titleCheck    = validateTitle(doc, page, pageLabel)
    val elementChecks = validatePageElements(doc, page, pageLabel)
    titleCheck :: elementChecks
  }

  def validateHtml(html: String, page: Page, pageLabel: String): List[ValidationResult] =
    validatePage(Jsoup.parse(html), page, pageLabel)

  private def validateTitle(doc: Document, page: Page, label: String): ValidationResult = {
    val h1Elements = doc.select("h1")
    if (h1Elements.isEmpty)
      Fail(s"$label — title", page.title, "<no h1 found>")
    else {
      val actualTitle = h1Elements.first().text().trim
      if (actualTitle == page.title) Pass(s"$label — title matches")
      else Fail(s"$label — title", s"'${page.title}'", s"'$actualTitle'")
    }
  }

  private def validatePageElements(doc: Document, page: Page, label: String): List[ValidationResult] =
    page.pageType match {
      case "string" =>
        val found = doc.select("input[type=text]").size()
        if (found >= 1) List(Pass(s"$label — has text input"))
        else List(Fail(s"$label — text input", "at least 1 <input type=text>", s"found $found"))

      case "datePage" =>
        val found = doc.select(
          "input[name*=day], input[name*=month], input[name*=year], input[type=date], .govuk-date-input"
        ).size()
        if (found >= 1) List(Pass(s"$label — has date fields"))
        else List(Fail(s"$label — date fields", "date input elements", s"found $found"))

      case "boolean" =>
        val found = doc.select("input[type=radio]").size()
        if (found >= 2) List(Pass(s"$label — has yes/no radios ($found found)"))
        else List(Fail(s"$label — boolean radios", "at least 2 radio buttons", s"found $found"))

      case "radioButton" =>
        val radios     = doc.select("input[type=radio]").size()
        val labelTexts = doc.select("label").asScala.map(_.text().trim).toSet

        val countCheck =
          if (radios >= page.options.size) List(Pass(s"$label — radio count ($radios found)"))
          else List(Fail(s"$label — radio count", s"${page.options.size} radio buttons", s"found $radios"))

        val optionChecks = page.options.map { opt =>
          if (labelTexts.exists(_.contains(opt))) Pass(s"$label — option '$opt' present")
          else Fail(s"$label — option '$opt'", s"label containing '$opt'", "not found")
        }

        countCheck ++ optionChecks

      case "checkbox" =>
        val found = doc.select("input[type=checkbox]").size()
        val countCheck =
          if (found >= page.options.size) List(Pass(s"$label — checkbox count ($found found)"))
          else List(Fail(s"$label — checkbox count", s"${page.options.size} checkboxes", s"found $found"))

        val labelTexts = doc.select("label").asScala.map(_.text().trim).toSet
        val optionChecks = page.options.map { opt =>
          if (labelTexts.exists(_.contains(opt))) Pass(s"$label — option '$opt' present")
          else Fail(s"$label — option '$opt'", s"label containing '$opt'", "not found")
        }

        countCheck ++ optionChecks

      case "multipleQuestionsPage" =>
        page.questions.map { q =>
          val found = doc.select("label, legend, .govuk-fieldset__heading").asScala
            .exists(_.text().trim.contains(q.questionTitle))
          if (found) Pass(s"$label — question '${q.questionTitle}' present")
          else Fail(s"$label — question", q.questionTitle, "not found on page")
        }

      case "contentPage" =>
        List(Pass(s"$label — content page (no form required)"))

      case other =>
        List(Fail(s"$label — page type", "known page type", s"'$other'"))
    }
}
